File: Uva11057.java
import java.util.*;
import java.io.*;

public class Uva11057
{

    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new FileReader("in.txt"));
        PrintWriter pw = new PrintWriter(System.out);
        StringBuilder sb = new StringBuilder();
        while( br.ready() ){
            int n = Integer.parseInt(br.readLine()) ;
            int [] a = new int[n] ;
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                a[i] = Integer.parseInt(st.nextToken()) ;
            }
            Arrays.sort(a);
            int m = Integer.parseInt(br.readLine()) ;
            int max_sum = 0 , book1 = 0 , book2 = 0 ;
            int min_diff = Integer.MAX_VALUE ;
            for(int i = 0 ; i < n ; i++){
                if( a[i] > m ){
                    break ;
                }
                int key = m - a[i] ;
                int idx = binary_search(a , key , i) ;
                if( idx == -1 ){
                    continue;
                }
                int sum = a[i] + a[idx] ;
                int diff = Integer.max(a[i],a[idx]) - Integer.min(a[i],a[idx]) ;
                if( sum > max_sum ){
                    max_sum = sum ;
                    min_diff = diff ;
                    book1 = a[i] ;
                    book2 = a[idx] ;
                } else if( sum == max_sum && diff < min_diff ){
                    min_diff = diff ;
                    book1 = a[i] ;
                    book2 = a[idx] ;
                }
            }
            sb.append("Peter should buy books whose prices are ").
                    append( Integer.min(book1,book2) ).
                    append(" and ").
                    append( Integer.max(book1,book2) ).
                    append(".\n\n");
            br.readLine() ;
        }
        pw.print(sb.toString());
        br.close();
        pw.close();
    }

    static int binary_search(int [] array , int key , int idx)
    {
        int low = 0 ;
        int high = array.length ;
        int x = 0 ;
        while( low <= high ){
            int mid = (low+high) >>> 1 ;
            int midVal = array[mid] ;
            if( midVal <= key && ( ( mid+1 < array.length && array[mid+1] > key ) || mid+1 == array.length ) ){
                if( mid == idx ){
                    return mid-1 ;
                } else {
                    return mid;
                }
            } else if( midVal > key ){
                high = mid-1 ;
            } else if( midVal < key ){
                low = mid+1 ;
            }
        }
        return -1 ;
    }

}

File: uva11057.cpp
/* @author : mostafa
 * created : 2018-06-30 20:20
 */
#include <iostream>
#include <algorithm>
#include <vector>
#include <climits>
using namespace std ;
#define forn(i,n) for(int i = 0 ; i < (int)n ; i++)

int main()
{
    cin.sync_with_stdio(false) ;
    cin.tie(0) ;
    int n ;
    while( cin >> n ){
        vector<int> a(n) ;
        forn(i,n){
            cin >> a[i] ;
        }
        sort( a.begin() , a.end() );
        int m ;
        cin >> m ;
        bool twice = m%2==0 ? true : false ;
        if ( twice ){
            int key = m/2 ;
            auto it = lower_bound( a.begin() , a.end() , key );
            int p = it - a.begin() ;
            if ( a[p] == key && a[p+1] == key  ){
                twice = true ;
            } else {
                twice = false ;
            }
        }
        int book1 = 0 , book2 = 0 ;
        int min_diff = INT_MAX ;
        int max_sum = 0 ;
        for(int i = 0 ; i < n ; i++){
            if ( a[i] > m ){
                break ;
            }
            int key = m - a[i] ;
            auto it = lower_bound( a.begin() , a.end() , key );
            int pos = it - a.begin() ;
            while( a[pos] < key ){
                pos-- ;
            }
            if ( pos == i  ){
                pos-- ;
            }
            if ( pos < 0 ){
                continue ;
            }
            int sum = a[i] + a[pos];
            int diff = max(a[i],a[pos]);
            
            if ( sum <= m && sum > max_sum ){
                max_sum = sum ; 
                min_diff = diff ;
                book1 = a[i] ;
                book2 = a[pos] ;
            } else if ( sum <= m && sum == max_sum ){
                if ( diff < min_diff ){
                    min_diff = diff ;
                    book1 = a[i] ;
                    book2 = a[pos] ;
                }
            }
        }
        int mn = min(book1,book2);
        int mx = max(book1,book2);
        cout << "Peter should buy books whose prices are " << mn << " and "<< mx <<".\n\n" ;
        string str ;
        getline( cin , str );
    }
    return 0 ;
}


